src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1367 @@
+        mdnsClientStarted_ = true;
+        notify();
+        LOG(INFO, "wifi: mdns_scan() start.");
+    }
+    else if (!mdnsClientTrigRefresh_)
+    {
+        // Trigger the mDNS state machine to execute early. If a query is
+        // already taking place, a new one will start immediately following it.
+        // If a query it not taking place, one will be started immediately.
+        mdnsClientTrigRefresh_ = true;
+        CallbackExecutable *e =
+            new CallbackExecutable([this](){this->timer_.ensure_triggered();});
+        this->service()->executor()->add(e);
+        LOG(INFO, "wifi: mdns_scan() trigger.");
+    }
+}
+
+//
+// EspIdfWiFiBase::wifi_event_handler()
+//
+void EspIdfWiFiBase::wifi_event_handler(
+    esp_event_base_t base, int32_t id, void *data)
+{
+    HASSERT(base == WIFI_EVENT);
+
+    switch (id)
+    {
+        default:
+            LOG(INFO, "wifi: wifi_event_handler() unknown id: %i", (int)id);
+            break;
+        case WIFI_EVENT_WIFI_READY:
+            started_ = true;
+            LOG(INFO, "wifi: WiFi started.");
+            break;
+        case WIFI_EVENT_STA_START:
+        {
+            std::string ssid;
+            std::string pass;
+            uint8_t authmode;
+            uint8_t channel;
+            last_sta_get(&ssid, &pass, &authmode, &channel);
+            if (ssid.empty())
+            {
+                // No valid "fast" connect parameters.
+                esp_wifi_scan_start(&SCAN_CONFIG, false);
+                LOG(INFO, "wifi: STA start, scanning...");
+            }
+            else
+            {
+                // Try the fast connect parameters first.
+                sta_connect(ssid, pass, authmode, channel);
+#if (LOGLEVEL == VERBOSE)
+                LOG(VERBOSE,
+                    "wifi: STA start, ssid: %s, pass: %s, connecting...",
+                    ssid.c_str(), pass.c_str());
+#else
+                LOG(INFO, "wifi: STA start, ssid: %s, connecting...",
+                    ssid.c_str());
+#endif
+            }
+            break;
+        }
+        case WIFI_EVENT_STA_CONNECTED:
+        {
+            HASSERT(!connected_);
+            connected_ = true;
+            wifi_event_sta_connected_t *sdata =
+                (wifi_event_sta_connected_t*)data;
+            LOG(INFO, "wifi: STA connected, channel: %u, authmode: %u",
+                sdata->channel, (unsigned)sdata->authmode);
+            // There is a very small chance for a race condition whereby another
+            // call "sta_connect()" call has been made with a different password
+            // after this connection has been established. The chances are very
+            // small and the harm is very limited to the next fast connect
+            // attempt failing.
+            last_sta_update(std::string((char*)sdata->ssid, sdata->ssid_len),
+                staConnectPass_, sdata->authmode, sdata->channel);
+            std::string ssid((char*)sdata->ssid, sdata->ssid_len);
```

Comment:
you could move this one line up as you have duplicate code now.

Reply:
Good catch.